File: backend/src/api/projects.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from src.api.deps import get_current_user, get_owned_project
from src.database.models import Project, User
from src.database.session import get_db
from src.models.project import ProjectCreate, ProjectOut
from src.services import project_service

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.post("", response_model=ProjectOut, status_code=status.HTTP_201_CREATED)
def create_project(
    payload: ProjectCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return project_service.create_project(db, current_user.id, payload.project_name)


@router.get("", response_model=list[ProjectOut])
def list_projects(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return project_service.list_projects(db, current_user.id)


@router.get("/{project_id}", response_model=ProjectOut)
def get_project(project: Project = Depends(get_owned_project)):
    return project

File: frontend/src/api/projects.ts
import { apiClient } from "./client";
import type { Project } from "../types/api";

export function listProjects() {
  return apiClient.get<Project[]>("/api/projects").then((r) => r.data);
}

export function createProject(projectName: string) {
  return apiClient
    .post<Project>("/api/projects", { project_name: projectName })
    .then((r) => r.data);
}

export function getProject(projectId: number) {
  return apiClient.get<Project>(`/api/projects/${projectId}`).then((r) => r.data);
}
